File: rt-desktop/src/main/java/xyz/redtorch/desktop/layout/base/bean/TradeFXBean.java
package xyz.redtorch.desktop.layout.base.bean;

import javafx.beans.property.SimpleIntegerProperty;
import javafx.beans.property.SimpleObjectProperty;
import javafx.beans.property.SimpleStringProperty;
import javafx.scene.layout.Pane;
import javafx.scene.layout.VBox;
import javafx.scene.text.Text;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import xyz.redtorch.common.util.CommonUtils;
import xyz.redtorch.pb.CoreEnum.DirectionEnum;
import xyz.redtorch.pb.CoreEnum.HedgeFlagEnum;
import xyz.redtorch.pb.CoreEnum.OffsetFlagEnum;
import xyz.redtorch.pb.CoreField.TradeField;

public class TradeFXBean {

    private final static Logger logger = LoggerFactory.getLogger(TradeFXBean.class);

    private final SimpleStringProperty tradeId = new SimpleStringProperty();
    private final SimpleObjectProperty<Pane> contract = new SimpleObjectProperty<>();
    private final SimpleObjectProperty<Text> direction = new SimpleObjectProperty<>();
    private final SimpleStringProperty offsetFlag = new SimpleStringProperty();
    private final SimpleStringProperty hedgeFlag = new SimpleStringProperty();
    private final SimpleStringProperty price = new SimpleStringProperty();
    private final SimpleIntegerProperty volume = new SimpleIntegerProperty();
    private final SimpleStringProperty tradeTime = new SimpleStringProperty();
    private final SimpleStringProperty adapterTradeId = new SimpleStringProperty();
    private final SimpleStringProperty originOrderId = new SimpleStringProperty();
    private final SimpleStringProperty accountId = new SimpleStringProperty();

    private TradeField tradeField;
    private boolean contractSelectedFlag = false;

    public TradeFXBean(TradeField tradeField, boolean contractSelectedFlag) {
        update(tradeField, contractSelectedFlag);
    }

    public TradeField getTradeField() {
        return tradeField;
    }

    public void update(TradeField newTradeField, boolean newContractSelectedFlag) {
        if (newTradeField == null) {
            return;
        }

        if (tradeField == null) {
            updateTradeId(newTradeField);
            updateDirection(newTradeField);
            updateOffsetFlag(newTradeField);
            updateHedgeFlag(newTradeField);
            updateContract(newTradeField);
            updateTradeTime(newTradeField);
            updateAdapterTradeId(newTradeField);
            updateAccountId(newTradeField);
        }
        if (contractSelectedFlag != newContractSelectedFlag) {
            contractSelectedFlag = newContractSelectedFlag;
            updateContract(newTradeField);
        }

        if (newTradeField != tradeField) {
            updateChangeable(newTradeField);
            tradeField = newTradeField;
        }
    }

    private void updateChangeable(TradeField newTradeField) {
        updatePrice(newTradeField);
        updateVolume(newTradeField);
        updateOriginOrderId(newTradeField);
    }

    private void updateTradeId(TradeField newTradeField) {
        setTradeId(newTradeField.getTradeId());
    }

    private void updateContract(TradeField newTradeField) {
        VBox vBox = new VBox();
        Text unifiedSymbolText = new Text(newTradeField.getContract().getUnifiedSymbol());
        Text shortNameText = new Text(newTradeField.getContract().getName());
        vBox.getChildren().add(unifiedSymbolText);
        vBox.getChildren().add(shortNameText);
        if (contractSelectedFlag) {
            unifiedSymbolText.getStyleClass().add("trade-remind-color");
        }
        vBox.setUserData(newTradeField);
        setContract(vBox);
    }

    private void updateDirection(TradeField newTradeField) {
        Text directionText = new Text();

        if (newTradeField.getDirection() == DirectionEnum.D_Buy) {
            directionText.setText("多");
            directionText.getStyleClass().add("trade-long-color");
        } else if (newTradeField.getDirection() == DirectionEnum.D_Sell) {
            directionText.setText("空");
            directionText.getStyleClass().add("trade-short-color");
        } else if (newTradeField.getDirection() == DirectionEnum.D_Unknown) {
            directionText.setText("未知");
        } else {
            directionText.setText(newTradeField.getDirection().getValueDescriptor().getName());
        }

        directionText.setUserData(newTradeField);

        setDirection(directionText);
    }

    private void updateOffsetFlag(TradeField newTradeField) {
        String offsetFlag;

        if (newTradeField.getOffsetFlag() == OffsetFlagEnum.OF_Close) {
            offsetFlag = "平";
        } else if (newTradeField.getOffsetFlag() == OffsetFlagEnum.OF_CloseToday) {
            offsetFlag = "平今";
        } else if (newTradeField.getOffsetFlag() == OffsetFlagEnum.OF_CloseYesterday) {
            offsetFlag = "平昨";
        } else if (newTradeField.getOffsetFlag() == OffsetFlagEnum.OF_Open) {
            offsetFlag = "开";
        } else if (newTradeField.getOffsetFlag() == OffsetFlagEnum.OF_Unknown) {
            offsetFlag = "未知";
        } else {
            offsetFlag = newTradeField.getOffsetFlag().getValueDescriptor().getName();
        }

        setOffsetFlag(offsetFlag);
    }

    private void updateHedgeFlag(TradeField newTradeField) {
        String hedgeFlag;

        if (newTradeField.getHedgeFlag() == HedgeFlagEnum.HF_Speculation) {
            hedgeFlag = "投机";
        } else if (newTradeField.getHedgeFlag() == HedgeFlagEnum.HF_Hedge) {
            hedgeFlag = "套保";
        } else if (newTradeField.getHedgeFlag() == HedgeFlagEnum.HF_Arbitrage) {
            hedgeFlag = "套利";
        } else if (newTradeField.getHedgeFlag() == HedgeFlagEnum.HF_MarketMaker) {
            hedgeFlag = "做市商";
        } else if (newTradeField.getHedgeFlag() == HedgeFlagEnum.HF_SpecHedge) {
            hedgeFlag = "第一条腿投机第二条腿套保 大商所专用";
        } else if (newTradeField.getHedgeFlag() == HedgeFlagEnum.HF_HedgeSpec) {
            hedgeFlag = "第一条腿套保第二条腿投机 大商所专用";
        } else if (newTradeField.getHedgeFlag() == HedgeFlagEnum.HF_Unknown) {
            hedgeFlag = "未知";
        } else {
            hedgeFlag = newTradeField.getHedgeFlag().getValueDescriptor().getName();
        }

        setHedgeFlag(hedgeFlag);
    }

    private void updatePrice(TradeField newTradeField) {
        if (tradeField == null || !CommonUtils.isEquals(tradeField.getPrice(),newTradeField.getPrice())) {
            String price = "渲染错误";
            try {
                int decimalDigits = CommonUtils.getNumberDecimalDigits(newTradeField.getContract().getPriceTick());
                if (decimalDigits < 0) {
                    decimalDigits = 0;
                }
                String priceStringFormat = "%,." + decimalDigits + "f";

                price = String.format(priceStringFormat, newTradeField.getPrice());
            } catch (Exception e) {
                logger.error("渲染错误", e);
            }
            setPrice(price);
        }
    }

    private void updateVolume(TradeField newTradeField) {
        if(tradeField == null || tradeField.getVolume() != newTradeField.getVolume()){
            setVolume(newTradeField.getVolume());
        }
    }

    private void updateTradeTime(TradeField newTradeField) {
        setTradeTime(newTradeField.getTradeTime());
    }

    private void updateOriginOrderId(TradeField newTradeField) {
        if (tradeField == null || !this.tradeField.getOriginOrderId().equals(newTradeField.getOriginOrderId())) {
            setOriginOrderId(newTradeField.getOriginOrderId());
        }
    }

    private void updateAdapterTradeId(TradeField newTradeField) {
        setAdapterTradeId(newTradeField.getAdapterTradeId());
    }

    private void updateAccountId(TradeField newTradeField) {
        setAccountId(newTradeField.getAccountId());
    }

    public String getTradeId() {
        return tradeId.get();
    }

    public void setTradeId(String tradeId) {
        this.tradeId.set(tradeId);
    }

    public SimpleStringProperty tradeIdProperty() {
        return tradeId;
    }

    public Pane getContract() {
        return contract.get();
    }

    public void setContract(Pane contract) {
        this.contract.set(contract);
    }

    public SimpleObjectProperty<Pane> contractProperty() {
        return contract;
    }

    public Text getDirection() {
        return direction.get();
    }

    public void setDirection(Text direction) {
        this.direction.set(direction);
    }

    public SimpleObjectProperty<Text> directionProperty() {
        return direction;
    }

    public String getOffsetFlag() {
        return offsetFlag.get();
    }

    public void setOffsetFlag(String offsetFlag) {
        this.offsetFlag.set(offsetFlag);
    }

    public SimpleStringProperty offsetFlagProperty() {
        return offsetFlag;
    }

    public String getHedgeFlag() {
        return hedgeFlag.get();
    }

    public void setHedgeFlag(String hedgeFlag) {
        this.hedgeFlag.set(hedgeFlag);
    }

    public SimpleStringProperty hedgeFlagProperty() {
        return hedgeFlag;
    }

    public String getPrice() {
        return price.get();
    }

    public void setPrice(String price) {
        this.price.set(price);
    }

    public SimpleStringProperty priceProperty() {
        return price;
    }

    public int getVolume() {
        return volume.get();
    }

    public void setVolume(int volume) {
        this.volume.set(volume);
    }

    public SimpleIntegerProperty volumeProperty() {
        return volume;
    }

    public String getTradeTime() {
        return tradeTime.get();
    }

    public void setTradeTime(String tradeTime) {
        this.tradeTime.set(tradeTime);
    }

    public SimpleStringProperty tradeTimeProperty() {
        return tradeTime;
    }

    public String getAdapterTradeId() {
        return adapterTradeId.get();
    }

    public void setAdapterTradeId(String adapterTradeId) {
        this.adapterTradeId.set(adapterTradeId);
    }

    public SimpleStringProperty adapterTradeIdProperty() {
        return adapterTradeId;
    }

    public String getOriginOrderId() {
        return originOrderId.get();
    }

    public void setOriginOrderId(String originOrderId) {
        this.originOrderId.set(originOrderId);
    }

    public SimpleStringProperty originOrderIdProperty() {
        return originOrderId;
    }

    public String getAccountId() {
        return accountId.get();
    }

    public void setAccountId(String accountId) {
        this.accountId.set(accountId);
    }

    public SimpleStringProperty accountIdProperty() {
        return accountId;
    }
}

File: rt-desktop/src/main/kotlin/xyz/redtorch/desktop/gui/bean/TradeFXBean.kt
package xyz.redtorch.desktop.gui.bean

import javafx.beans.property.SimpleIntegerProperty
import javafx.beans.property.SimpleObjectProperty
import javafx.beans.property.SimpleStringProperty
import javafx.scene.layout.Pane
import javafx.scene.layout.VBox
import javafx.scene.text.Text
import xyz.redtorch.common.trade.dto.Trade
import xyz.redtorch.common.trade.enumeration.DirectionEnum
import xyz.redtorch.common.trade.enumeration.HedgeFlagEnum
import xyz.redtorch.common.trade.enumeration.OffsetFlagEnum
import xyz.redtorch.common.utils.CommonUtils

class TradeFXBean(trade: Trade, contractSelectedFlag: Boolean) {

    private val tradeId = SimpleStringProperty()
    private val contract = SimpleObjectProperty<Pane>()
    private val direction = SimpleObjectProperty<Text>()
    private val offsetFlag = SimpleStringProperty()
    private val hedgeFlag = SimpleStringProperty()
    private val price = SimpleStringProperty()
    private val volume = SimpleIntegerProperty()
    private val tradeTime = SimpleStringProperty()
    private val adapterTradeId = SimpleStringProperty()
    private val originOrderId = SimpleStringProperty()
    private val accountId = SimpleStringProperty()
    private var decimalDigits = 4
    private var contractSelectedFlag = false
    var tradeDTO: Trade? = null
        private set

    init {
        update(trade, contractSelectedFlag)
    }

    fun update(newTradeDTO: Trade, newContractSelectedFlag: Boolean) {
        if (tradeDTO == null) {
            decimalDigits = CommonUtils.getNumberDecimalDigits(newTradeDTO.contract.priceTick)
            updateTradeId(newTradeDTO)
            updateDirection(newTradeDTO)
            updateOffsetFlag(newTradeDTO)
            updateHedgeFlag(newTradeDTO)
            updateContract(newTradeDTO)
            updateTradeTime(newTradeDTO)
            updateAdapterTradeId(newTradeDTO)
            updateAccountId(newTradeDTO)
        }
        if (contractSelectedFlag != newContractSelectedFlag) {
            contractSelectedFlag = newContractSelectedFlag
            updateContract(newTradeDTO)
        }
        // 对比内存地址
        if (newTradeDTO !== tradeDTO) {
            updateChangeable(newTradeDTO)
            tradeDTO = newTradeDTO
        }
    }

    private fun updateChangeable(newTradeDTO: Trade) {
        updatePrice(newTradeDTO)
        updateVolume(newTradeDTO)
        updateOriginOrderId(newTradeDTO)
    }

    private fun updateTradeId(newTradeDTO: Trade) {
        setTradeId(newTradeDTO.tradeId)
    }

    private fun updateContract(newTradeDTO: Trade) {
        val vBox = VBox().apply {
            children.add(Text(newTradeDTO.contract.uniformSymbol).apply {
                // 选中合约高亮
                if (contractSelectedFlag) {
                    styleClass.add("trade-remind-color")
                }
            })
            children.add(
                Text(newTradeDTO.contract.name)
            )
            userData = newTradeDTO
        }
        setContract(vBox)
    }

    private fun updateDirection(newTradeDTO: Trade) {
        val directionText = Text()
        when (newTradeDTO.direction) {
            DirectionEnum.Buy -> {
                directionText.text = "多"
                directionText.styleClass.add("trade-long-color")
            }
            DirectionEnum.Sell -> {
                directionText.text = "空"
                directionText.styleClass.add("trade-short-color")
            }
            DirectionEnum.Unknown -> {
                directionText.text = "未知"
            }
            else -> {
                directionText.text = newTradeDTO.direction.toString()
            }
        }

        directionText.userData = newTradeDTO
        setDirection(directionText)
    }

    private fun updateOffsetFlag(newTradeDTO: Trade) {
        val offsetFlag = when (newTradeDTO.offsetFlag) {
            OffsetFlagEnum.Close -> {
                "平"
            }
            OffsetFlagEnum.CloseToday -> {
                "平今"
            }
            OffsetFlagEnum.CloseYesterday -> {
                "平昨"
            }
            OffsetFlagEnum.Open -> {
                "开"
            }
            OffsetFlagEnum.Unknown -> {
                "未知"
            }
            else -> {
                newTradeDTO.offsetFlag.toString()
            }
        }
        setOffsetFlag(offsetFlag)
    }

    private fun updateHedgeFlag(newTradeDTO: Trade) {
        val hedgeFlag = when (newTradeDTO.hedgeFlag) {
            HedgeFlagEnum.Speculation -> {
                "投机"
            }
            HedgeFlagEnum.Hedge -> {
                "套保"
            }
            HedgeFlagEnum.Arbitrage -> {
                "套利"
            }
            HedgeFlagEnum.MarketMaker -> {
                "做市商"
            }
            HedgeFlagEnum.SpecHedge -> {
                "第一条腿投机第二条腿套保 大商所专用"
            }
            HedgeFlagEnum.HedgeSpec -> {
                "第一条腿套保第二条腿投机 大商所专用"
            }
            HedgeFlagEnum.Unknown -> {
                "未知"
            }
            else -> {
                newTradeDTO.hedgeFlag.toString()
            }
        }
        setHedgeFlag(hedgeFlag)
    }

    private fun updatePrice(newTradeDTO: Trade) {
        if (tradeDTO == null || CommonUtils.isNotEquals(tradeDTO!!.price, newTradeDTO.price)) {
            setPrice(CommonUtils.formatDouble(newTradeDTO.price, decimalDigits))
        }
    }

    private fun updateVolume(newTradeDTO: Trade) {
        if (tradeDTO == null || tradeDTO!!.volume != newTradeDTO.volume) {
            setVolume(newTradeDTO.volume)
        }
    }

    private fun updateTradeTime(newTradeDTO: Trade) {
        setTradeTime(newTradeDTO.tradeTime)
    }

    private fun updateOriginOrderId(newTradeDTO: Trade) {
        if (tradeDTO == null || tradeDTO!!.originOrderId != newTradeDTO.originOrderId) {
            setOriginOrderId(newTradeDTO.originOrderId)
        }
    }

    private fun updateAdapterTradeId(newTradeDTO: Trade) {
        setAdapterTradeId(newTradeDTO.adapterTradeId)
    }

    private fun updateAccountId(newTradeDTO: Trade) {
        setAccountId(newTradeDTO.accountId)
    }

    fun getTradeId(): String {
        return tradeId.get()
    }

    fun setTradeId(tradeId: String?) {
        this.tradeId.set(tradeId)
    }

    fun tradeIdProperty(): SimpleStringProperty {
        return tradeId
    }

    fun getContract(): Pane {
        return contract.get()
    }

    fun setContract(contract: Pane) {
        this.contract.set(contract)
    }

    fun contractProperty(): SimpleObjectProperty<Pane> {
        return contract
    }

    fun getDirection(): Text {
        return direction.get()
    }

    fun setDirection(direction: Text) {
        this.direction.set(direction)
    }

    fun directionProperty(): SimpleObjectProperty<Text> {
        return direction
    }

    fun getOffsetFlag(): String {
        return offsetFlag.get()
    }

    fun setOffsetFlag(offsetFlag: String?) {
        this.offsetFlag.set(offsetFlag)
    }

    fun offsetFlagProperty(): SimpleStringProperty {
        return offsetFlag
    }

    fun getHedgeFlag(): String {
        return hedgeFlag.get()
    }

    fun setHedgeFlag(hedgeFlag: String?) {
        this.hedgeFlag.set(hedgeFlag)
    }

    fun hedgeFlagProperty(): SimpleStringProperty {
        return hedgeFlag
    }

    fun getPrice(): String {
        return price.get()
    }

    fun setPrice(price: String?) {
        this.price.set(price)
    }

    fun priceProperty(): SimpleStringProperty {
        return price
    }

    fun getVolume(): Int {
        return volume.get()
    }

    fun setVolume(volume: Int) {
        this.volume.set(volume)
    }

    fun volumeProperty(): SimpleIntegerProperty {
        return volume
    }

    fun getTradeTime(): String {
        return tradeTime.get()
    }

    fun setTradeTime(tradeTime: String?) {
        this.tradeTime.set(tradeTime)
    }

    fun tradeTimeProperty(): SimpleStringProperty {
        return tradeTime
    }

    fun getAdapterTradeId(): String {
        return adapterTradeId.get()
    }

    fun setAdapterTradeId(adapterTradeId: String?) {
        this.adapterTradeId.set(adapterTradeId)
    }

    fun adapterTradeIdProperty(): SimpleStringProperty {
        return adapterTradeId
    }

    fun getOriginOrderId(): String {
        return originOrderId.get()
    }

    fun setOriginOrderId(originOrderId: String?) {
        this.originOrderId.set(originOrderId)
    }

    fun originOrderIdProperty(): SimpleStringProperty {
        return originOrderId
    }

    fun getAccountId(): String {
        return accountId.get()
    }

    fun setAccountId(accountId: String?) {
        this.accountId.set(accountId)
    }

    fun accountIdProperty(): SimpleStringProperty {
        return accountId
    }
}